File: lista3/zad2/DHUser.java
//package zad2;

import java.util.Random;

public class DHUser<T extends Operations<T>> {
    private T privateKey;
    private boolean isKeySet = false;
    private long secretPower;
    private DHSetup<T> dhs;

    public DHUser(DHSetup<T> dhs, int characteristics) {
        this.dhs = dhs;
        Random rand = new Random();
        secretPower = rand.nextInt(characteristics - 1);
    }

    public T getPublicKey() {
        return dhs.power(dhs.getGenerator(), secretPower);
    }

    public void setKey(T a) {
        privateKey = dhs.power(a, secretPower);
        isKeySet = true;
    }

    public T encrypt(T m) throws Exception {
        if (!isKeySet) {
            throw new Exception("You have not set the private key yet.");
        }
        return m.multiply(privateKey.getValue());
    }

    public T decrypt(T c) throws Exception {
        if (!isKeySet) {
            throw new Exception("You have not set the private key yet.");
        }
        return c.divide(privateKey.getValue());
    }
}

File: lista3/zad3/DHUser.py

#from typing import TypeVar
import random
import copy
from diffieHellman import DHSetup, T

class User:

    secretPower: int

    def __init__(self, dhs: DHSetup, characteristics: int):
        self.secretPower = random.randint(1, characteristics - 1)
        self.dhs = dhs
        self.privateKey: T = None

    def getPublicKey(self) -> T:
        b = copy.copy(self.secretPower)
        return self.dhs.power(self.dhs.getGenerator(), b)

    def setKey(self, a: T):
        b = copy.copy(self.secretPower)
        self.privateKey = self.dhs.power(a, b)

    def encrypt(self, m: T) -> T:
        if self.privateKey == None:
            raise Exception("You have not set the private key yet.")
        return m * self.privateKey

    def decrypt(self, c: T) -> T:
        if self.privateKey == None:
            raise Exception("You have not set the private key yet.")
        return c / self.privateKey